x-pack/elastic-agent/pkg/core/monitoring/server/process.go

```diff
@@ -0,0 +1,180 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package server
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/gorilla/mux"
+
+	"github.com/elastic/beats/v7/metricbeat/mb/parse"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/program"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/artifact"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/core/monitoring/beats"
+)
+
+const (
+	processIDKey      = "processID"
+	monitoringSuffix  = "-monitoring"
+	separator         = "-"
+	timeout           = 10 * time.Second
+	errTypeUnexpected = "UNEXPECTED"
+
+	httpPlusPrefix = "http+"
+)
+
+var (
+	// ErrProgramNotSupported returned when requesting metrics for not supported program.
+	ErrProgramNotSupported = errors.New("specified program is not supported")
+)
+
+func processHandler() func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Content-Type", "application/json; charset=utf-8")
+
+		vars := mux.Vars(r)
+		id, found := vars[processIDKey]
+
+		if !found {
+			writeResponse(
+				w,
+				unexpectedErrorWithReason("productID not found"),
+			)
+			return
+		}
+
+		metricsBytes, statusCode, metricsErr := processMetrics(r.Context(), id)
+		switch metricsErr {
+		case ErrProgramNotSupported:
+			w.WriteHeader(http.StatusNotFound)
+		default:
+			w.WriteHeader(statusCode)
+		}
+
+		if metricsErr != nil {
+			writeResponse(w, unexpectedErrorWithReason("failed fetching metrics: %s", metricsErr.Error()))
+			return
+		}
+
+		fmt.Fprint(w, string(metricsBytes))
```

Comment:
I think I normally like is that metrics endpoint are also human readable. What I mean in the context here is that we pretty print the json. Unfortunately this means in this context to convert it first to json to be able to pretty print it with indentation. At the same time, should not cause too much overhead?

Reply:
how do you specify pretty print with metricbeat? we can pass argument to mb if passed to agent

Reply:
As far as I remember there is a `?pretty` flag: `/stats/?pretty`. Not sure if that works over the socket. I was initially thinking to implement it here so it works for all the outputs also agent, but not strong preference.

Reply:
I think the pretty was a special flag provided by the expvar handler. Not sure we have had it implemented in Beats.

Reply:
BTW this is not a blocker, please ignore it for now.